src/python/library/tritonclient/utils/cuda_shared_memory/cuda_shared_memory.cc

```diff
@@ -128,58 +147,73 @@ CudaSharedMemoryGetRawHandle(
 
 int
 CudaSharedMemoryRegionSet(
-    void* cuda_shm_handle, size_t offset, size_t byte_size, const void* data)
+    void* cuda_shm_handle, size_t offset, size_t byte_size, const void* data,
+    int device_id)
 {
-  // remember previous device and set to new device
-  int previous_device;
-  cudaGetDevice(&previous_device);
-  cudaError_t err = cudaSetDevice(
-      reinterpret_cast<SharedMemoryHandle*>(cuda_shm_handle)->device_id_);
-  if (err != cudaSuccess) {
-    cudaSetDevice(previous_device);
-    return -1;
+  auto lhandle = reinterpret_cast<SharedMemoryHandle*>(cuda_shm_handle);
+
+  {
+    // Unified virtual addressing is the prerequisite for cudaMemcpyDefault,
+    // unsupported platform is rare so only for sanity check.
+    auto res = SupportUVA(lhandle->device_id_, device_id);
+    if (res != 0) {
+      return res;
+    }
   }
 
   // Copy data into cuda shared memory
-  void* base_addr =
-      reinterpret_cast<SharedMemoryHandle*>(cuda_shm_handle)->base_addr_;
-  err = cudaMemcpy(
+  void* base_addr = lhandle->base_addr_;
+  cudaError_t err = cudaMemcpy(
       reinterpret_cast<uint8_t*>(base_addr) + offset, data, byte_size,
-      cudaMemcpyHostToDevice);
+      cudaMemcpyDefault);
   if (err != cudaSuccess) {
-    cudaSetDevice(previous_device);
     return -3;
   }
 
-  // Set device to previous GPU
-  cudaSetDevice(previous_device);
```

Comment:
I think removing the set devices would close: https://github.com/triton-inference-server/server/issues/5715